controllers/machine_controller.go

```diff
@@ -181,6 +182,37 @@ func (r *MachineReconciler) reconcile(ctx context.Context, cluster *clusterv1.Cl
 	return res, kerrors.NewAggregate(errs)
 }
 
+func (r *MachineReconciler) reconcileLables(ctx context.Context, m *clusterv1.Machine) error {
+	logger := r.Log.WithValues("machine", m.Name, "namespace", m.Namespace)
+
+	machineControllerRef := metav1.GetControllerOf(m)
+	if machineControllerRef == nil {
+		return nil
+	}
+	if m.Labels == nil {
+		m.Labels = make(map[string]string)
+	}
+	m.Labels[clusterv1.MachineSetLabelName] = machineControllerRef.Name
+
+	ms := &clusterv1.MachineSet{}
+	err := r.Client.Get(ctx, client.ObjectKey{
+		Namespace: m.Namespace,
+		Name:      machineControllerRef.Name,
+	}, ms)
+	if err != nil {
+		logger.Error(err, "failed to get owner machineset", "machineset", machineControllerRef.Name)
+		return err
+	}
+
+	machineSetControllerRef := metav1.GetControllerOf(ms)
+	if machineSetControllerRef == nil {
+		return nil
+	}
+	m.Labels[clusterv1.MachineDeploymentLabelName] = machineSetControllerRef.Name
```

Comment:
This could be simplified like this
```golang
	if md := metav1.GetControllerOf(ms); md != nil {
		m.Labels[clusterv1.MachineDeploymentLabelName] = md.Name
	}
```

Reply:
thanks @vincepri ! I've updated the pr, PTAL :)